vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go

```diff
@@ -46,109 +50,227 @@ type PodMetricsLister interface {
+	return fmt.Sprintf("max_over_time(container_memory_rss{%s!='', %s=~'%s', namespace='%s'}[5m])", containerNameLabel, podNameLabel, strings.Join(podNames, "|"), namespace)
 }
 
-func getJVMHeapCommittedQuery(containerName string, podName string, namespace string) string {
-	return fmt.Sprintf("max_over_time(jmx_Memory_HeapMemoryUsage_committed{kubernetes_container_name='%s', kubernetes_pod_name='%s', kubernetes_namespace='%s'}[5m])", containerName, podName, namespace)
+// getJVMHeapCommittedQuery returns a query string for the JVM Heap Committed metric for the specified pods in the namespace.
+func getJVMHeapCommittedQuery(podNames []string, namespace string, containerNameLabel string, podNameLabel string) string {
+	return fmt.Sprintf("max_over_time(jmx_Memory_HeapMemoryUsage_committed{%s!='', %s=~'%s', kubernetes_namespace='%s'}[5m])", containerNameLabel, podNameLabel, strings.Join(podNames, "|"), namespace)
 }
 
-// Augments the PodMetricsList with custom metrics from M3.
-func (s podMetricsSource) withM3CustomMetrics(podMetrics *v1beta1.PodMetricsList) (*v1beta1.PodMetricsList, error) {
-	m3Url, err := url.Parse(s.m3Url)
+// queryCustomResourceMetric queries M3 for a custom resource metric for the specified pods in the namespace.
+func (s podMetricsSource) queryCustomResourceMetric(ns string, podNames []string, customQuery nsQuery) nsQueryResults {
+	params := url.Values{}
+	params.Add("query", customQuery.f(podNames, ns, customQuery.containerNameLabel, customQuery.podNameLabel))
+	s.m3Url.RawQuery = params.Encode()
+
+	resp, err := http.Get(s.m3Url.String())
 	if err != nil {
-		klog.ErrorS(err, "Failed to parse M3 URL")
-		return nil, err
+		return nsQueryResults{namespace: ns, resource: customQuery.resource, err: err}
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nsQueryResults{namespace: ns, resource: customQuery.resource, err: fmt.Errorf("Failed to get valid response (status: %s)", resp.Status)}
 	}
-	m3Url.Path += "/api/v1/query"
 
-	for i, pod := range podMetrics.Items {
-		for j, container := range pod.Containers {
-			queries := map[string]k8sapiv1.ResourceName{
-				getRSSQuery(container.Name, pod.Name, pod.Namespace):              k8sapiv1.ResourceName(model.ResourceRSS),
-				getJVMHeapCommittedQuery(container.Name, pod.Name, pod.Namespace): k8sapiv1.ResourceName(model.ResourceJVMHeapCommitted),
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nsQueryResults{namespace: ns, resource: customQuery.resource, err: err}
+	}
+
+	var response m3Response
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nsQueryResults{namespace: ns, resource: customQuery.resource, err: err}
+	}
+
+	nsQueryResults := nsQueryResults{
+		namespace: ns,
+		resource:  customQuery.resource,
+		podUsages: make(map[string]containerUsages),
+	}
+	for _, result := range response.Data.Result {
```

Comment:
(I missed to parse the logic.. ) does it throw errors if the pod/container does not have jvm metric (in legit cases, service is not scala service)?

Reply:
No, it ignores this at the moment